Python/chapter02/2.5 - Sum Lists/miguel_2.5_soln.py

```diff
@@ -149,96 +149,25 @@ def sum_lists_forward(ll1: LinkedList, ll2: LinkedList) -> LinkedList:
+    reversed_ll2 = LinkedList()
+    while n1 is not None:
+        reversed_ll1.append_to_head(n1.data)
+        n1 = n1.next
+    while n2 is not None:
+        reversed_ll2.append_to_head(n2.data)
+        n2 = n2.next
+    # then, call sum_lists
+    reversed_result = sum_lists(reversed_ll1, reversed_ll2)
+    # reverse a final time
     output_ll = LinkedList()
-    # perform addition, taking into account a
-    # 'backwards' carry value since we
-    # are looping forward digits down to
-    # 'ones' digit place
-    prev_node = None
-    leading_bw_carry = False
-    while n_shorter is not None:
-        first = n_shorter.data
-        second = n_longer.data
-        result = first + second
-        if result >= 10:
-            backwards_carry = 1
-            result -= 10  # extract digit in one's place
-            if prev_node is None:
-                # for when the first addition yields a carry
-                # and the output linked list is empty
-                leading_bw_carry = True
-            else:
-                # add carry to previous digit
-                prev_node.data += backwards_carry
-        output_ll.append_to_tail(result)
-        n_shorter = n_shorter.next
-        n_longer = n_longer.next
-        prev_node = output_ll.tail
-
-    # first case after adding same digit places,
-    # both input lists same size
-    if length_diff == 0:
-        if leading_bw_carry:
-            backwards_carry = 1
-            output_ll.append_to_head(backwards_carry)
-        return output_ll
-    # otherwise, we have non-equal lengths
-    # with or without a leading backward carry
-    # leading bw carry means the the size diff
-    # decreased.
-    # no leading bw carry means size diff stayed the same
-    if leading_bw_carry:
-        backwards_carry = 1
-    else:
-        backwards_carry = 0
-    # otherwise, as long as the diff is >= 0
-    # we want to take care of the leading digits.
-    # start by decrementing the diff
-    length_diff -= 1
-    while length_diff >= 0:
-        # advance the n_longer pointer from longer_ll head
-        # up to where the current diff is.
-        n_longer = longer_ll.head
-        for i in range(length_diff):
-            n_longer = n_longer.next
-
-        if backwards_carry == 1:
-            output_ll.append_to_head(backwards_carry)
-            # output data will have the backwards_carry
-            # appended from before
-            result = output_ll.head.data + n_longer.data
-            if result >= 10:
-                result -= 10
-            else:
-                backwards_carry = 0
-            output_ll.head.data = result
-        else:
-            # if backwards carry is 0, bring corresponding digit down
-            result = n_longer.data
-            output_ll.append_to_head(result)
-        length_diff -= 1
-    # last backwards carry
-    if backwards_carry == 1:
-        output_ll.append_to_head(backwards_carry)
+    n = reversed_result.head
```

Comment:
i'm a bit on the fence about this extra copy. there's a way to reverse a linked list in O(1) space, but it's a bit more complex. and linked lists are sort of boring, so up to you to create an additional version of the linked list reversing function that destructively reverses the output list (you can add it to the LinkedList class and can do it "in-place").

Reply:
updated